File: HRMS/HRMS.Web/Modules/ServerTypes/Communication/SocialPostForm.ts
import { DateEditor, EnumEditor, initFormType, LookupEditor, PrefixedContext, StringEditor, TextAreaEditor, Widget } from "@serenity-is/corelib";
import { SocialPostStatus } from "./SocialPostStatus";
import { SocialPostType } from "./SocialPostType";

export interface SocialPostForm {
    SocialAccountId: LookupEditor;
    Title: StringEditor;
    Content: TextAreaEditor;
    PostType: EnumEditor;
    Status: EnumEditor;
    ScheduledDate: DateEditor;
    PublishedDate: DateEditor;
    MediaList: Widget;
}

export class SocialPostForm extends PrefixedContext {
    static readonly formKey = 'Communication.SocialPost';
    private static init: boolean;

    constructor(prefix: string) {
        super(prefix);

        if (!SocialPostForm.init) {
            SocialPostForm.init = true;

            var w0 = LookupEditor;
            var w1 = StringEditor;
            var w2 = TextAreaEditor;
            var w3 = EnumEditor;
            var w4 = DateEditor;
            var w5 = Widget;

            initFormType(SocialPostForm, [
                'SocialAccountId', w0,
                'Title', w1,
                'Content', w2,
                'PostType', w3,
                'Status', w3,
                'ScheduledDate', w4,
                'PublishedDate', w4,
                'MediaList', w5
            ]);
        }
    }
}

[SocialPostType, SocialPostStatus]; // referenced types

File: HRMS/HRMS.Web/Modules/Communication/SocialPost/SocialPostForm.cs
using Serenity.ComponentModel;

namespace HRMS.Communication.Forms;

[FormScript("Communication.SocialPost")]
[BasedOnRow(typeof(SocialPostRow), CheckNames = true)]
public class SocialPostForm
{
    public int SocialAccountId { get; set; }
    public string Title { get; set; }
    [TextAreaEditor(Rows = 6)]
    public string Content { get; set; }
    public SocialPostType PostType { get; set; }
    public SocialPostStatus Status { get; set; }
    public DateTime ScheduledDate { get; set; }
    public DateTime PublishedDate { get; set; }
    [SocialPostMediaEditorAttribute]
    public List<SocialPostMediaRow> MediaList { get; set; }
}

public class SocialPostMediaEditorAttribute : CustomEditorAttribute
{
    public const string Key = "HRMS.Communication.SocialPostMediaEditor";

    public SocialPostMediaEditorAttribute()
        : base(Key)
    {
    }
}
